rclcpp/src/rclcpp/publisher.cpp

```diff
@@ -231,6 +230,19 @@ PublisherBase::setup_intra_process(
   IntraProcessManagerSharedPtr ipm,
   const rcl_publisher_options_t & intra_process_options)
 {
+  // Skip intraprocess configuration if the "durability" qos policy is not "volatile".
+  // TODO(ivanpauno):
+  //    RMW_QOS_POLICY_DURABILITY_SYSTEM_DEFAULT is for FastRTPS and Connext also
+  //    equal to "volatile". It would be good to add a
+  //    is_durability_policy_volatile(rcl_publisher_options_t options)
+  //    method to rcl, and corresponging methods in rmw and the rmw vendor
+  //    implementations, which will also checks if "system_default" is "volatile" or not.
+  if (intra_process_options.qos.durability != RMW_QOS_POLICY_DURABILITY_VOLATILE) {
+    RCLCPP_WARN(
+      rclcpp::get_logger("rclcpp"),
+      "Enabled intraprocess communication with incompatible QoS policy.");
```

Comment:
I think the warning should be more specific about the consequence:

> Skipping intraprocess communication with incompatible QoS policy.

Reply:
@ivanpauno are you not planning to add the publisher/subscription option to enable intra-process and therefore make this an error? If you are planning to do so, then I think this warning does not need to be reworded as it will be removed.

Reply:
I will do it in another PR. For the moment, I will left the warning reworded.

Reply:
I don't really agree, I think warning isn't really sufficient, as I said in my other comment this basically prevents you from using intra-process and transient-local durability at the same time (even if on different entities). So after merging this pr we'll be potentially breaking existing use cases for a very narrow optimization. I would prefer to see the improvement done completely in this pr than have two pr's.